File: cpp/vhal/client/src/AidlHalPropConfig.cpp
/*
 * Copyright (c) 2022, The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "AidlHalPropConfig.h"

#include <VehicleUtils.h>

namespace android {
namespace frameworks {
namespace automotive {
namespace vhal {

using ::aidl::android::hardware::automotive::vehicle::VehicleAreaConfig;
using ::aidl::android::hardware::automotive::vehicle::VehiclePropConfig;
using ::aidl::android::hardware::automotive::vehicle::VehiclePropertyAccess;

using ::android::hardware::automotive::vehicle::toInt;

AidlHalPropConfig::AidlHalPropConfig(VehiclePropConfig&& config) {
    mPropConfig = std::move(config);
    if (mPropConfig.areaConfigs.size() == 0) {
        VehicleAreaConfig globalAreaConfig;
        globalAreaConfig.areaId = 0;
        mAreaConfigs.push_back(std::make_unique<AidlHalAreaConfig>(std::move(globalAreaConfig),
                                                                   toInt(mPropConfig.access)));
    } else {
        for (VehicleAreaConfig& areaConfig : mPropConfig.areaConfigs) {
            int32_t access = (areaConfig.access == VehiclePropertyAccess::NONE)
                    ? toInt(mPropConfig.access)
                    : toInt(areaConfig.access);
            mAreaConfigs.push_back(
                    std::make_unique<AidlHalAreaConfig>(std::move(areaConfig), access));
        }
    }
}

int32_t AidlHalPropConfig::getPropId() const {
    return mPropConfig.prop;
}

int32_t AidlHalPropConfig::getAccess() const {
    return toInt(mPropConfig.access);
}

int32_t AidlHalPropConfig::getChangeMode() const {
    return toInt(mPropConfig.changeMode);
}

size_t AidlHalPropConfig::getAreaConfigSize() const {
    return mAreaConfigs.size();
}

std::vector<int32_t> AidlHalPropConfig::getConfigArray() const {
    return mPropConfig.configArray;
}

std::string AidlHalPropConfig::getConfigString() const {
    return mPropConfig.configString;
}

float AidlHalPropConfig::getMinSampleRate() const {
    return mPropConfig.minSampleRate;
}

float AidlHalPropConfig::getMaxSampleRate() const {
    return mPropConfig.maxSampleRate;
}

AidlHalAreaConfig::AidlHalAreaConfig(VehicleAreaConfig&& areaConfig, int32_t access) {
    mAreaConfig = std::move(areaConfig);
    mAccess = access;
}

int32_t AidlHalAreaConfig::getAreaId() const {
    return mAreaConfig.areaId;
}

int32_t AidlHalAreaConfig::getAccess() const {
    return mAccess;
}

int32_t AidlHalAreaConfig::getMinInt32Value() const {
    return mAreaConfig.minInt32Value;
}

int32_t AidlHalAreaConfig::getMaxInt32Value() const {
    return mAreaConfig.maxInt32Value;
}

int64_t AidlHalAreaConfig::getMinInt64Value() const {
    return mAreaConfig.minInt64Value;
}

int64_t AidlHalAreaConfig::getMaxInt64Value() const {
    return mAreaConfig.maxInt64Value;
}

float AidlHalAreaConfig::getMinFloatValue() const {
    return mAreaConfig.minFloatValue;
}

float AidlHalAreaConfig::getMaxFloatValue() const {
    return mAreaConfig.maxFloatValue;
}

bool AidlHalAreaConfig::isVariableUpdateRateSupported() const {
    return mAreaConfig.supportVariableUpdateRate;
}

}  // namespace vhal
}  // namespace automotive
}  // namespace frameworks
}  // namespace android

File: service/src/com/android/car/hal/AidlHalPropConfig.java
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.car.hal;

import static com.android.car.internal.common.CommonConstants.EMPTY_INT_ARRAY;
import static com.android.car.internal.ExcludeFromCodeCoverageGeneratedReport.DUMP_INFO;

import android.hardware.automotive.vehicle.VehicleAreaConfig;
import android.hardware.automotive.vehicle.VehiclePropConfig;
import android.hardware.automotive.vehicle.VehiclePropertyAccess;

import com.android.car.internal.ExcludeFromCodeCoverageGeneratedReport;

/**
 * AidlHalPropConfig is a HalPropConfig with an AIDL backend.
 */
public final class AidlHalPropConfig extends HalPropConfig {
    private static final VehicleAreaConfig[] sEmptyAreaConfig = new VehicleAreaConfig[0];

    private final VehiclePropConfig mConfig;
    private final HalAreaConfig[] mAreaConfigs;

    public AidlHalPropConfig(VehiclePropConfig config) {
        mConfig = config;

        // Do some validity checks to make sure we do not return null for get functions.
        if (mConfig.areaConfigs == null) {
            mConfig.areaConfigs = sEmptyAreaConfig;
        }
        if (mConfig.configString == null) {
            mConfig.configString = "";
        }
        if (mConfig.configArray == null) {
            mConfig.configArray = EMPTY_INT_ARRAY;
        }

        int size = mConfig.areaConfigs.length;
        mAreaConfigs = new HalAreaConfig[size];
        for (int i = 0; i < size; i++) {
            if (mConfig.areaConfigs[i].access == VehiclePropertyAccess.NONE) {
                mConfig.areaConfigs[i].access = mConfig.access;
            }
            mAreaConfigs[i] = new AidlHalAreaConfig(mConfig.areaConfigs[i]);
        }
    }

    /**
     * Get the property ID.
     */
    @Override
    public int getPropId() {
        return mConfig.prop;
    }

    /**
     * Get the access mode.
     */
    @Override
    public int getAccess() {
        return mConfig.access;
    }

    /**
     * Get the change mode.
     */
    @Override
    public int getChangeMode() {
        return mConfig.changeMode;
    }

    /**
     * Get the area configs.
     */
    @Override
    public HalAreaConfig[] getAreaConfigs() {
        return mAreaConfigs;
    }

    /**
     * Get the config array.
     */
    @Override
    public int[] getConfigArray() {
        return mConfig.configArray;
    }

    /**
     * Get the config string.
     */
    @Override
    public String getConfigString() {
        return mConfig.configString;
    }

    /**
     * Get the min sample rate.
     */
    @Override
    public float getMinSampleRate() {
        return mConfig.minSampleRate;
    }

    /**
     * Get the max sample rate.
     */
    @Override
    public float getMaxSampleRate() {
        return mConfig.maxSampleRate;
    }

    /**
     * Converts to AIDL or HIDL VehiclePropConfig.
     */
    @Override
    public Object toVehiclePropConfig() {
        return mConfig;
    }

    /**
     * Get the string representation for debugging.
     */
    @Override
    @ExcludeFromCodeCoverageGeneratedReport(reason = DUMP_INFO)
    public String toString() {
        return mConfig.toString();
    }
}
